### ISSUE TITLE
Having trouble including other nunjucks template files

### ISSUE THREAD

**Issue by thepechinator** [at0g/nunjucks-loader#7]
Feels like I'm missing something obvious here, but I can't get {% include %} to work. I tried it in the paths-examples directory:

I added a 'test.nunj' file under path/to/templates with contents like

``` html_example
blah
```

Then in path/to/templates/child.nunj, I added the {% include %} like so:

``` html_example
{% block content %}
<div style="background-color: #ffff00; padding: 1em;">
    <pre><code>path/to/templates/child.nunj</code></pre>
    <h2>Child message: {{ message | default('(none)') }}</h2>
</div>

{% include 'test.nunj' %}

{% endblock %}
```

Webpack bundles everything fine, but when I hit the page I get an error in the console:

Uncaught Template render error: (unknown path)
  Template render error: Template render error: TypeError: Cannot read property 'render' of undefined

Any ideas? Also, just fyi, `npm run test` all pass for me. Thanks in advance.

**Comment by at0g** [comment]
This is definitely a bug - thanks for reporting it.

I'm currently getting errors from both node/web tests using the 1.3.x versions of nunjucks, but the 1.1.x/1.2.x versions are working ok - when you get a chance, can you please let me know which version of nunjucks you have installed.

With the example you gave above, I've been trying some permutations of include and found the following:

* include within a block (fails)
* include a file outside of a block (works as expected)
* include a file that contains a block (works as expected)

I'm hoping to have some time on Friday to investigate (and hopefully resolve) the issue.

**Comment by thepechinator** [comment]
Thanks for the quick response.

I forgot to take out changes based on https://github.com/at0g/nunjucks-loader/pull/5, which was causing all the web tests to pass with the nunjucks version I was using (1.3.4).

With a fresh checkout, I'm seeing the same thing with the web tests failing with nunjucks 1.3.x.

**Comment by thepechinator** [comment]
I've been playing around with things today, and I think I think I figured out part of the problem and one solution. There were two issues: 

1) The newest nunjucks (1.3.4) changed the number of parameters by 1 for Environment.getTemplate so it looks something like Environment.getTemplate( name, ec, parentName, cb ). I updated env.getTemplate in runtime-shim.js to match this and it resolved all of the web test errors. This also made it so all the includes worked again, except for the case of it being within a block.

2) I think in the case of the {% include %} in a block, it's treating it as another require call, which causes another pass through index.js and call to runtime-shim with an empty dependencies object.  The source of the issue which causes the error we're seeing in the console is in runtime-shim.js: 

```javascript
env.getTemplate = function( name, ec, cb ) {
    if( typeof ec === "function" ) {
        cb = ec;
        ec = false;
    }
    var _require = function(name) {
        try {
            // It's looking for dependencies['path/to/templates/test.nunj'] here, which doesn't exist.
            // It does exist in the pass that happens with the parent.nunj template though.
            return dependencies[name];
        }
        catch (e) {
            if ( frame.get( "_require" ) ) return frame.get( "_require" )( name )
        }
    };

    // The tmpl isn't found in the dependencies object and no exception is thrown.
    // Eventually something tries to do tmpl.compile() and it throws the error we see in the console.
    var tmpl = _require( name );
```

I'm not sure whether this is what you want, but I solved it by keeping track of the dependencies object from previous calls to index.js. What I ended up with that works in all cases with the web tests passing:

**runtime-shim.js**
```javascript
module.exports = function(nunjucks, env, obj, dependencies){

    var oldRoot = obj.root;
    obj.root = function( env, context, frame, runtime, cb ) {
        var oldGetTemplate = env.getTemplate;

        // Just updated the number of parameters to match latest nunjucks
        env.getTemplate = function( name, ec, parentName, cb ) {
            if( typeof ec === "function" ) {
                cb = ec;
                ec = false;
            }
            var _require = function(name) {
                try {
                    // add a reference to the already resolved dependency here...
                    return dependencies[name];
                }
                catch (e) {
                    if ( frame.get( "_require" ) ) return frame.get( "_require" )( name )
                }
            };

            var tmpl = _require( name );
            frame.set( "_require", _require );
            if( ec ) tmpl.compile();
            cb( null, tmpl );
        };


        oldRoot( env, context, frame, runtime, function( err, res ) {
            env.getTemplate = oldGetTemplate;
            cb( err, res );
        });
    };

    var src = {
        obj: obj,
        type: 'code'
    };

    return new nunjucks.Template( src, env );

};
```

**index.js** (Only update is changing dependencies to window.dependencies. I just used window to get it working, but there is probably something better to use if you want to maintain the dependencies object through multiple passes)
```javascript
module.exports = function(source) {
    this.cacheable();

    if(!hasRun){
        var query = this.query.replace('?', '');
        if(query.length > 0){
            var q = JSON.parse(query);

            if(q.config){
                pathToConfigure = q.config;
                var configure = require(q.config);
                configure(env);
            }
        }
        hasRun = true;
    }

    var nunjucksCompiledStr = compiler.compile(source, env.asyncFilters, env.extensionsList);
    var reg = /env\.getTemplate\(\"(.*?)\"/g;
    var match;
    var required = {};
    var compiledTemplate = '';

    if (this.target === 'web') {
        compiledTemplate += 'var nunjucks = require( "nunjucks/browser/nunjucks-slim" );\n';
    }
    else {
        compiledTemplate += 'var nunjucks = require("nunjucks");\n';
    }

    compiledTemplate += 'var env = new nunjucks.Environment([], {dev: true, autoescape: true});\n';
    // Add a global dependencies object to hold resolved dependencies and maintain for
    // consecutive calls. I just used window here because it was a quick way to get
    // things done. There is probably a better way to keep track of this if you want
    // to maintain it through multiple passes.
    compiledTemplate += 'window.dependencies = window.dependencies || {};\n';

    if( pathToConfigure ){
        compiledTemplate += 'var configure = require("' + slash(pathToConfigure) + '")(env);\n';
    }

    while( match = reg.exec( nunjucksCompiledStr ) ) {
        var templateRef = match[1];

        if (!required[templateRef]) {
            // Require the dependency by name, so it get bundled by webpack
            compiledTemplate += 'window.dependencies["' + templateRef + '"] = require( "' + templateRef + '" );\n';
            required[templateRef] = true;
        }
    }
    compiledTemplate += '\n\n\n\n';
    compiledTemplate += 'var shim = require("' + slash(__dirname + '/runtime-shim') + '");\n';
    compiledTemplate += 'var obj = (function () {' + nunjucksCompiledStr + '})();\n';
    compiledTemplate += 'module.exports = shim(nunjucks, env, obj, window.dependencies)';

    return compiledTemplate;
};
```

Hope this helps. Please let me know if you need me to clarify anything.

**Comment by at0g** [comment]
@thepechinator - Thanks for the info, I appreciate your help resolving this.

I'm not 100% sold on the window/global approach, but it will do for now.

I've published a new version `1.0.5` to npm.
